File: skills/tooling/managing-marketplace/scripts/verify-marketplace.py
#!/usr/bin/env python3
"""Verify marketplace.json integrity and skill compliance.

Standalone script (stdlib only) for CI or local use.
Checks marketplace structure, skill paths, frontmatter, naming, and orphans.

Usage:
    python verify-marketplace.py [--json] [--fix] [marketplace.json]
"""

import argparse
import json
import os
import re
import sys
from pathlib import Path


# --- Constants ---

REQUIRED_MARKETPLACE_FIELDS = {"name", "plugins"}
REQUIRED_PLUGIN_FIELDS = {"name", "source"}
RECOMMENDED_PLUGIN_FIELDS = {"description", "category"}
REQUIRED_FRONTMATTER_FIELDS = {"name", "description"}
MAX_NAME_LENGTH = 64
MAX_DESCRIPTION_LENGTH = 1024
MAX_SKILL_LINES = 500
NAME_PATTERN = re.compile(r"^[a-z][a-z0-9]*(-[a-z0-9]+)*$")
GERUND_SUFFIXES = ("ing", "ment", "tion", "sion")
FIRST_PERSON_STARTS = ("i ", "we ", "my ", "our ")


# --- Report data ---

class Report:
    def __init__(self):
        self.errors = []
        self.warnings = []
        self.info = []
        self.stats = {
            "plugins_checked": 0,
            "skills_checked": 0,
            "skills_missing": 0,
            "orphans_found": 0,
        }

    def error(self, rule, message, path=None, fix=None):
        entry = {"severity": "error", "rule": rule, "message": message}
        if path:
            entry["path"] = str(path)
        if fix:
            entry["fix"] = fix
        self.errors.append(entry)

    def warning(self, rule, message, path=None, fix=None):
        entry = {"severity": "warning", "rule": rule, "message": message}
        if path:
            entry["path"] = str(path)
        if fix:
            entry["fix"] = fix
        self.warnings.append(entry)

    def passed(self):
        return len(self.errors) == 0

    def to_dict(self):
        return {
            "status": "PASS" if self.passed() else "FAIL",
            "errors": self.errors,
            "warnings": self.warnings,
            "stats": self.stats,
        }

    def to_table(self):
        lines = []
        status = "PASS" if self.passed() else "FAIL"
        lines.append(f"## Marketplace Validation Report")
        lines.append(f"")
        lines.append(f"Status: {status}")
        lines.append(f"")

        if self.errors:
            lines.append(f"### Errors ({len(self.errors)})")
            for e in self.errors:
                path_str = f" {e['path']}" if e.get("path") else ""
                fix_str = f"\n  Fix: {e['fix']}" if e.get("fix") else ""
                lines.append(f"- [{e['rule']}]{path_str} - {e['message']}{fix_str}")
            lines.append("")

        if self.warnings:
            lines.append(f"### Warnings ({len(self.warnings)})")
            for w in self.warnings:
                path_str = f" {w['path']}" if w.get("path") else ""
                fix_str = f"\n  Fix: {w['fix']}" if w.get("fix") else ""
                lines.append(f"- [{w['rule']}]{path_str} - {w['message']}{fix_str}")
            lines.append("")

        lines.append(f"### Summary")
        lines.append(f"- Plugins checked: {self.stats['plugins_checked']}")
        lines.append(f"- Skills checked: {self.stats['skills_checked']}")
        lines.append(f"- Skills missing: {self.stats['skills_missing']}")
        lines.append(f"- Orphans found: {self.stats['orphans_found']}")
        lines.append(f"- Errors: {len(self.errors)}")
        lines.append(f"- Warnings: {len(self.warnings)}")

        return "\n".join(lines)


# --- Frontmatter parser ---

def parse_frontmatter(content):
    """Parse YAML frontmatter from SKILL.md content. Returns (dict, error)."""
    lines = content.split("\n")
    if not lines or lines[0].strip() != "---":
        return None, "No YAML frontmatter found (missing opening ---)"

    end_idx = None
    for i in range(1, len(lines)):
        if lines[i].strip() == "---":
            end_idx = i
            break

    if end_idx is None:
        return None, "Unclosed YAML frontmatter (missing closing ---)"

    fm = {}
    current_key = None
    current_list = None

    for line in lines[1:end_idx]:
        stripped = line.strip()
        if not stripped or stripped.startswith("#"):
            continue

        # List item
        if stripped.startswith("- "):
            if current_key and current_list is not None:
                current_list.append(stripped[2:].strip())
            continue

        # Key-value pair
        if ":" in stripped:
            key, _, value = stripped.partition(":")
            key = key.strip()
            value = value.strip()

            # Save previous list if any
            if current_key and current_list is not None:
                fm[current_key] = current_list

            current_key = key
            if value:
                fm[key] = value
                current_list = None
            else:
                current_list = []

    # Save final list
    if current_key and current_list is not None:
        fm[current_key] = current_list

    return fm, None


# --- Validators ---

def validate_marketplace_structure(data, report):
    """Check marketplace.json top-level structure."""
    for field in REQUIRED_MARKETPLACE_FIELDS:
        if field not in data:
            report.error("required-field", f"Missing required field: {field}")

    if "plugins" in data and not isinstance(data["plugins"], list):
        report.error("invalid-type", "Field 'plugins' must be an array")


def validate_plugin(plugin, idx, all_skill_paths, report, repo_root):
    """Validate a single plugin entry."""
    report.stats["plugins_checked"] += 1
    plugin_name = plugin.get("name", f"plugin[{idx}]")

    for field in REQUIRED_PLUGIN_FIELDS:
        if field not in plugin:
            report.error("required-field", f"Plugin '{plugin_name}': missing required field '{field}'")

    for field in RECOMMENDED_PLUGIN_FIELDS:
        if field not in plugin:
            report.warning("recommended-field", f"Plugin '{plugin_name}': missing recommended field '{field}'")

    skills = plugin.get("skills", [])
    if not isinstance(skills, list):
        report.error("invalid-type", f"Plugin '{plugin_name}': 'skills' must be an array")
        return

    for skill_path in skills:
        report.stats["skills_checked"] += 1
        normalized = skill_path.lstrip("./")
        skill_md = repo_root / normalized / "SKILL.md"

        if not skill_md.is_file():
            report.stats["skills_missing"] += 1
            report.error(
                "missing-skill",
                f"Skill path does not exist: {skill_path}",
                path=str(skill_md),
                fix=f"Remove '{skill_path}' from plugin '{plugin_name}' or create {skill_md}",
            )
            continue

        # Check for duplicate paths across plugins
        if normalized in all_skill_paths:
            report.error(
                "duplicate-path",
                f"Skill path '{skill_path}' appears in both '{all_skill_paths[normalized]}' and '{plugin_name}'",
            )
        all_skill_paths[normalized] = plugin_name

        # Validate SKILL.md content
        validate_skill_md(skill_md, normalized, report)


def validate_skill_md(skill_md, skill_path, report):
    """Validate individual SKILL.md file."""
    try:
        content = skill_md.read_text(encoding="utf-8")
    except Exception as e:
        report.error("read-error", f"Cannot read {skill_md}: {e}", path=str(skill_md))
        return

    line_count = len(content.split("\n"))
    if line_count > MAX_SKILL_LINES:
        report.warning(
            "skill-too-large",
            f"SKILL.md is {line_count} lines (max {MAX_SKILL_LINES})",
            path=str(skill_md),
        )

    fm, err = parse_frontmatter(content)
    if err:
        report.error("invalid-frontmatter", err, path=str(skill_md))
        return

    if fm is None:
        report.error("missing-frontmatter", "No frontmatter found", path=str(skill_md))
        return

    # Check required frontmatter fields
    for field in REQUIRED_FRONTMATTER_FIELDS:
        if field not in fm:
            report.error(
                "missing-frontmatter-field",
                f"Missing required frontmatter field: {field}",
                path=str(skill_md),
            )

    # Validate name
    name = fm.get("name", "")
    if name:
        validate_name(name, skill_path, skill_md, report)

    # Validate description
    desc = fm.get("description", "")
    if desc:
        validate_description(desc, skill_md, report)


def validate_name(name, skill_path, skill_md, report):
    """Validate skill name against naming conventions."""
    if len(name) > MAX_NAME_LENGTH:
        report.error(
            "name-too-long",
            f"Name '{name}' exceeds {MAX_NAME_LENGTH} chars ({len(name)})",
            path=str(skill_md),
        )

    if not NAME_PATTERN.match(name):
        report.error(
            "invalid-name-format",
            f"Name '{name}' must be lowercase letters, numbers, and hyphens only",
            path=str(skill_md),
        )

    if "--" in name:
        report.error(
            "consecutive-hyphens",
            f"Name '{name}' contains consecutive hyphens",
            path=str(skill_md),
        )

    # Check gerund form (first word should end in common gerund suffixes)
    first_word = name.split("-")[0]
    if not any(first_word.endswith(s) for s in GERUND_SUFFIXES):
        report.warning(
            "non-gerund-name",
            f"Name '{name}' first word '{first_word}' doesn't appear to be gerund form",
            path=str(skill_md),
        )

    # Check name matches directory (allow parent-dir pattern for namespacing)
    parts = Path(skill_path).parts
    dir_name = parts[-1] if parts else ""
    parent_dir = parts[-2] if len(parts) >= 2 else ""
    parent_prefixed = f"{parent_dir}-{dir_name}" if parent_dir else ""
    if name != dir_name and name != parent_prefixed:
        report.warning(
            "name-mismatch",
            f"Frontmatter name '{name}' doesn't match directory name '{dir_name}' or '{parent_prefixed}'",
            path=str(skill_md),
        )


def validate_description(desc, skill_md, report):
    """Validate skill description."""
    if len(desc) > MAX_DESCRIPTION_LENGTH:
        report.error(
            "description-too-long",
            f"Description exceeds {MAX_DESCRIPTION_LENGTH} chars ({len(desc)})",
            path=str(skill_md),
        )

    lower = desc.lower().strip()
    if any(lower.startswith(p) for p in FIRST_PERSON_STARTS):
        report.warning(
            "first-person-description",
            "Description should be third-person (avoid 'I', 'We')",
            path=str(skill_md),
        )

    if "use when" not in lower:
        report.warning(
            "missing-use-when",
            "Description should include 'Use when' trigger clause",
            path=str(skill_md),
        )


def find_orphan_skills(repo_root, registered_paths, report):
    """Find SKILL.md files on disk not registered in marketplace.json."""
    skills_dir = repo_root / "skills"
    if not skills_dir.is_dir():
        return []

    orphans = []
    for skill_md in skills_dir.rglob("SKILL.md"):
        rel = skill_md.parent.relative_to(repo_root)
        normalized = str(rel)
        if normalized not in registered_paths:
            orphans.append(normalized)
            report.stats["orphans_found"] += 1
            report.warning(
                "orphan-skill",
                f"Skill exists on disk but not in marketplace.json: {normalized}",
                path=str(skill_md),
                fix=f"Add './{normalized}' to a plugin in marketplace.json",
            )

    return orphans


def check_unique_plugin_names(plugins, report):
    """Check that plugin names are unique."""
    seen = {}
    for i, plugin in enumerate(plugins):
        name = plugin.get("name", "")
        if not name:
            continue
        if name in seen:
            report.error(
                "duplicate-plugin-name",
                f"Plugin name '{name}' is used by plugins at index {seen[name]} and {i}",
            )
        seen[name] = i


# --- Fix mode ---

def apply_fixes(marketplace_path, data, repo_root, report):
    """Apply auto-fixes: remove missing skills, add orphans."""
    fixed = False

    # Remove missing skill paths
    for plugin in data.get("plugins", []):
        skills = plugin.get("skills", [])
        valid = []
        for sp in skills:
            normalized = sp.lstrip("./")
            skill_md = repo_root / normalized / "SKILL.md"
            if skill_md.is_file():
                valid.append(sp)
            else:
                print(f"  Removed missing: {sp} from {plugin.get('name', '?')}")
                fixed = True
        plugin["skills"] = valid

    # Remove plugins with no skills
    original_count = len(data["plugins"])
    data["plugins"] = [p for p in data["plugins"] if p.get("skills")]
    if len(data["plugins"]) < original_count:
        print(f"  Removed {original_count - len(data['plugins'])} empty plugins")
        fixed = True

    if fixed:
        with open(marketplace_path, "w", encoding="utf-8") as f:
            json.dump(data, f, indent=2, ensure_ascii=False)
            f.write("\n")
        print(f"\n  Fixed marketplace.json written to {marketplace_path}")

    return fixed


# --- Main ---

def verify(marketplace_path, output_json=False, fix=False):
    """Run all validations and produce report."""
    report = Report()
    marketplace_path = Path(marketplace_path).resolve()
    repo_root = marketplace_path.parent.parent  # .claude-plugin/ -> repo root

    # Parse marketplace.json
    if not marketplace_path.is_file():
        report.error("file-not-found", f"File not found: {marketplace_path}")
        return report

    try:
        with open(marketplace_path, encoding="utf-8") as f:
            data = json.load(f)
    except json.JSONDecodeError as e:
        report.error("invalid-json", f"Invalid JSON: {e}")
        return report

    # Validate structure
    validate_marketplace_structure(data, report)

    plugins = data.get("plugins", [])
    if not isinstance(plugins, list):
        return report

    # Check unique names
    check_unique_plugin_names(plugins, report)

    # Validate each plugin and its skills
    all_skill_paths = {}
    for i, plugin in enumerate(plugins):
        validate_plugin(plugin, i, all_skill_paths, report, repo_root)

    # Find orphans
    find_orphan_skills(repo_root, all_skill_paths, report)

    # Apply fixes if requested
    if fix and not report.passed():
        apply_fixes(marketplace_path, data, repo_root, report)

    return report


def main():
    parser = argparse.ArgumentParser(description="Verify marketplace.json integrity")
    parser.add_argument("path", nargs="?", default=".claude-plugin/marketplace.json",
                        help="Path to marketplace.json (default: .claude-plugin/marketplace.json)")
    parser.add_argument("--json", action="store_true", help="Output as JSON")
    parser.add_argument("--fix", action="store_true",
                        help="Auto-fix: remove missing skills, remove empty plugins")
    args = parser.parse_args()

    report = verify(args.path, output_json=args.json, fix=args.fix)

    if args.json:
        print(json.dumps(report.to_dict(), indent=2))
    else:
        print(report.to_table())

    sys.exit(0 if report.passed() else 1)


if __name__ == "__main__":
    main()

File: scripts/verify-marketplace.ts
#!/usr/bin/env bun
/**
 * Verify marketplace.json integrity and skill compliance.
 * Checks marketplace structure, skill paths, frontmatter, naming, and orphans.
 *
 * Usage: bun run scripts/verify-marketplace.ts [--json] [--fix] [marketplace.json]
 */

import { readFileSync, writeFileSync, existsSync } from "node:fs";
import { resolve, relative, basename, dirname } from "node:path";
import { parseArgs } from "node:util";
import { globSync } from "node:fs";

// --- Constants ---

const REQUIRED_MARKETPLACE_FIELDS = ["name", "plugins"] as const;
const REQUIRED_PLUGIN_FIELDS = ["name", "source"] as const;
const RECOMMENDED_PLUGIN_FIELDS = ["description", "category"] as const;
const REQUIRED_FRONTMATTER_FIELDS = ["name", "description"] as const;
const MAX_NAME_LENGTH = 64;
const MAX_DESCRIPTION_LENGTH = 1024;
const MAX_SKILL_LINES = 500;
const NAME_PATTERN = /^[a-z][a-z0-9]*(-[a-z0-9]+)*$/;
const GERUND_SUFFIXES = ["ing", "ment", "tion", "sion"];
const FIRST_PERSON_STARTS = ["i ", "we ", "my ", "our "];

// --- Types ---

interface Entry {
  severity: "error" | "warning";
  rule: string;
  message: string;
  path?: string;
  fix?: string;
}

// --- Report ---

class Report {
  errors: Entry[] = [];
  warnings: Entry[] = [];
  stats = { plugins_checked: 0, skills_checked: 0, skills_missing: 0, orphans_found: 0 };

  error(rule: string, message: string, path?: string, fix?: string) {
    this.errors.push({ severity: "error", rule, message, ...(path && { path }), ...(fix && { fix }) });
  }
  warning(rule: string, message: string, path?: string, fix?: string) {
    this.warnings.push({ severity: "warning", rule, message, ...(path && { path }), ...(fix && { fix }) });
  }
  passed() {
    return this.errors.length === 0;
  }

  toDict() {
    return { status: this.passed() ? "PASS" : "FAIL", errors: this.errors, warnings: this.warnings, stats: this.stats };
  }

  toTable() {
    const lines: string[] = [];
    lines.push("## Marketplace Validation Report", "", `Status: ${this.passed() ? "PASS" : "FAIL"}`, "");

    if (this.errors.length) {
      lines.push(`### Errors (${this.errors.length})`);
      for (const e of this.errors) {
        const p = e.path ? ` ${e.path}` : "";
        const f = e.fix ? `\n  Fix: ${e.fix}` : "";
        lines.push(`- [${e.rule}]${p} - ${e.message}${f}`);
      }
      lines.push("");
    }

    if (this.warnings.length) {
      lines.push(`### Warnings (${this.warnings.length})`);
      for (const w of this.warnings) {
        const p = w.path ? ` ${w.path}` : "";
        const f = w.fix ? `\n  Fix: ${w.fix}` : "";
        lines.push(`- [${w.rule}]${p} - ${w.message}${f}`);
      }
      lines.push("");
    }

    lines.push(
      "### Summary",
      `- Plugins checked: ${this.stats.plugins_checked}`,
      `- Skills checked: ${this.stats.skills_checked}`,
      `- Skills missing: ${this.stats.skills_missing}`,
      `- Orphans found: ${this.stats.orphans_found}`,
      `- Errors: ${this.errors.length}`,
      `- Warnings: ${this.warnings.length}`,
    );
    return lines.join("\n");
  }
}

// --- Frontmatter parser ---

function parseFrontmatter(content: string): { data: Record<string, string | string[]> | null; error: string | null } {
  const lines = content.split("\n");
  if (!lines.length || lines[0].trim() !== "---") return { data: null, error: "No YAML frontmatter found (missing opening ---)" };

  let endIdx: number | null = null;
  for (let i = 1; i < lines.length; i++) {
    if (lines[i].trim() === "---") { endIdx = i; break; }
  }
  if (endIdx === null) return { data: null, error: "Unclosed YAML frontmatter (missing closing ---)" };

  const fm: Record<string, string | string[]> = {};
  let currentKey: string | null = null;
  let currentList: string[] | null = null;

  for (const line of lines.slice(1, endIdx)) {
    const stripped = line.trim();
    if (!stripped || stripped.startsWith("#")) continue;

    if (stripped.startsWith("- ") && currentKey && currentList !== null) {
      currentList.push(stripped.slice(2).trim());
      continue;
    }

    const colonIdx = stripped.indexOf(":");
    if (colonIdx !== -1) {
      if (currentKey && currentList !== null) fm[currentKey] = currentList;
      const key = stripped.slice(0, colonIdx).trim();
      const value = stripped.slice(colonIdx + 1).trim();
      currentKey = key;
      if (value) { fm[key] = value; currentList = null; }
      else currentList = [];
    }
  }
  if (currentKey && currentList !== null) fm[currentKey] = currentList;

  return { data: fm, error: null };
}

// --- Validators ---

function validateName(name: string, skillPath: string, skillMdPath: string, report: Report) {
  if (name.length > MAX_NAME_LENGTH) report.error("name-too-long", `Name '${name}' exceeds ${MAX_NAME_LENGTH} chars (${name.length})`, skillMdPath);
  if (!NAME_PATTERN.test(name)) report.error("invalid-name-format", `Name '${name}' must be lowercase letters, numbers, and hyphens only`, skillMdPath);
  if (name.includes("--")) report.error("consecutive-hyphens", `Name '${name}' contains consecutive hyphens`, skillMdPath);

  const firstWord = name.split("-")[0];
  if (!GERUND_SUFFIXES.some((s) => firstWord.endsWith(s))) {
    report.warning("non-gerund-name", `Name '${name}' first word '${firstWord}' doesn't appear to be gerund form`, skillMdPath);
  }

  const parts = skillPath.split("/");
  const dirName = parts.at(-1) ?? "";
  const parentDir = parts.length >= 2 ? parts.at(-2) ?? "" : "";
  const parentPrefixed = parentDir ? `${parentDir}-${dirName}` : "";
  if (name !== dirName && name !== parentPrefixed) {
    report.warning("name-mismatch", `Frontmatter name '${name}' doesn't match directory name '${dirName}' or '${parentPrefixed}'`, skillMdPath);
  }
}

function validateDescription(desc: string, skillMdPath: string, report: Report) {
  if (desc.length > MAX_DESCRIPTION_LENGTH) report.error("description-too-long", `Description exceeds ${MAX_DESCRIPTION_LENGTH} chars (${desc.length})`, skillMdPath);
  const lower = desc.toLowerCase().trim();
  if (FIRST_PERSON_STARTS.some((p) => lower.startsWith(p))) report.warning("first-person-description", "Description should be third-person (avoid 'I', 'We')", skillMdPath);
  if (!lower.includes("use when")) report.warning("missing-use-when", "Description should include 'Use when' trigger clause", skillMdPath);
}

function validateSkillMd(skillMdPath: string, skillPath: string, report: Report) {
  let content: string;
  try {
    content = readFileSync(skillMdPath, "utf-8");
  } catch (e) {
    report.error("read-error", `Cannot read ${skillMdPath}: ${e}`, skillMdPath);
    return;
  }

  const lineCount = content.split("\n").length;
  if (lineCount > MAX_SKILL_LINES) report.warning("skill-too-large", `SKILL.md is ${lineCount} lines (max ${MAX_SKILL_LINES})`, skillMdPath);

  const { data: fm, error: err } = parseFrontmatter(content);
  if (err) { report.error("invalid-frontmatter", err, skillMdPath); return; }
  if (!fm) { report.error("missing-frontmatter", "No frontmatter found", skillMdPath); return; }

  for (const field of REQUIRED_FRONTMATTER_FIELDS) {
    if (!(field in fm)) report.error("missing-frontmatter-field", `Missing required frontmatter field: ${field}`, skillMdPath);
  }

  const name = fm.name;
  if (typeof name === "string" && name) validateName(name, skillPath, skillMdPath, report);
  const desc = fm.description;
  if (typeof desc === "string" && desc) validateDescription(desc, skillMdPath, report);
}

function validatePlugin(
  plugin: Record<string, unknown>,
  idx: number,
  allSkillPaths: Map<string, string>,
  report: Report,
  repoRoot: string,
) {
  report.stats.plugins_checked++;
  const pluginName = (plugin.name as string) || `plugin[${idx}]`;

  for (const field of REQUIRED_PLUGIN_FIELDS) {
    if (!(field in plugin)) report.error("required-field", `Plugin '${pluginName}': missing required field '${field}'`);
  }
  for (const field of RECOMMENDED_PLUGIN_FIELDS) {
    if (!(field in plugin)) report.warning("recommended-field", `Plugin '${pluginName}': missing recommended field '${field}'`);
  }

  const skills = plugin.skills;
  if (!Array.isArray(skills)) {
    report.error("invalid-type", `Plugin '${pluginName}': 'skills' must be an array`);
    return;
  }

  for (const skillPath of skills as string[]) {
    report.stats.skills_checked++;
    const normalized = skillPath.replace(/^\.\//, "");
    const skillMd = resolve(repoRoot, normalized, "SKILL.md");

    if (!existsSync(skillMd)) {
      report.stats.skills_missing++;
      report.error("missing-skill", `Skill path does not exist: ${skillPath}`, skillMd, `Remove '${skillPath}' from plugin '${pluginName}' or create ${skillMd}`);
      continue;
    }

    if (allSkillPaths.has(normalized)) {
      report.error("duplicate-path", `Skill path '${skillPath}' appears in both '${allSkillPaths.get(normalized)}' and '${pluginName}'`);
    }
    allSkillPaths.set(normalized, pluginName);
    validateSkillMd(skillMd, normalized, report);
  }
}

function findOrphanSkills(repoRoot: string, registeredPaths: Map<string, string>, report: Report) {
  const skillsDir = resolve(repoRoot, "skills");
  if (!existsSync(skillsDir)) return;

  // Use Bun.glob for finding SKILL.md files
  const glob = new Bun.Glob("**/SKILL.md");
  for (const match of glob.scanSync({ cwd: skillsDir })) {
    const skillDir = dirname(match);
    const normalized = `skills/${skillDir}`;
    if (!registeredPaths.has(normalized)) {
      report.stats.orphans_found++;
      report.warning("orphan-skill", `Skill exists on disk but not in marketplace.json: ${normalized}`, resolve(skillsDir, match), `Add './${normalized}' to a plugin in marketplace.json`);
    }
  }
}

function applyFixes(marketplacePath: string, data: Record<string, unknown>, repoRoot: string) {
  let fixed = false;
  const plugins = data.plugins as Array<Record<string, unknown>>;

  for (const plugin of plugins) {
    const skills = plugin.skills as string[];
    const valid = skills.filter((sp) => {
      const normalized = sp.replace(/^\.\//, "");
      const skillMd = resolve(repoRoot, normalized, "SKILL.md");
      if (existsSync(skillMd)) return true;
      console.log(`  Removed missing: ${sp} from ${plugin.name ?? "?"}`);
      fixed = true;
      return false;
    });
    plugin.skills = valid;
  }

  const originalCount = plugins.length;
  data.plugins = plugins.filter((p) => (p.skills as string[]).length > 0);
  if ((data.plugins as unknown[]).length < originalCount) {
    console.log(`  Removed ${originalCount - (data.plugins as unknown[]).length} empty plugins`);
    fixed = true;
  }

  if (fixed) {
    writeFileSync(marketplacePath, JSON.stringify(data, null, 2) + "\n");
    console.log(`\n  Fixed marketplace.json written to ${marketplacePath}`);
  }
  return fixed;
}

// --- Main ---

function verify(marketplacePath: string, fix: boolean): Report {
  const report = new Report();
  const resolved = resolve(marketplacePath);
  const repoRoot = resolve(dirname(resolved), "..");

  if (!existsSync(resolved)) {
    report.error("file-not-found", `File not found: ${resolved}`);
    return report;
  }

  let data: Record<string, unknown>;
  try {
    data = JSON.parse(readFileSync(resolved, "utf-8"));
  } catch (e) {
    report.error("invalid-json", `Invalid JSON: ${e}`);
    return report;
  }

  for (const field of REQUIRED_MARKETPLACE_FIELDS) {
    if (!(field in data)) report.error("required-field", `Missing required field: ${field}`);
  }
  if ("plugins" in data && !Array.isArray(data.plugins)) {
    report.error("invalid-type", "Field 'plugins' must be an array");
  }

  const plugins = data.plugins;
  if (!Array.isArray(plugins)) return report;

  // Check unique names
  const seen = new Map<string, number>();
  for (let i = 0; i < plugins.length; i++) {
    const name = plugins[i].name as string;
    if (!name) continue;
    if (seen.has(name)) report.error("duplicate-plugin-name", `Plugin name '${name}' is used by plugins at index ${seen.get(name)} and ${i}`);
    seen.set(name, i);
  }

  const allSkillPaths = new Map<string, string>();
  for (let i = 0; i < plugins.length; i++) {
    validatePlugin(plugins[i], i, allSkillPaths, report, repoRoot);
  }

  findOrphanSkills(repoRoot, allSkillPaths, report);

  if (fix && !report.passed()) applyFixes(resolved, data, repoRoot);

  return report;
}

const { values, positionals } = parseArgs({
  args: process.argv.slice(2),
  options: {
    json: { type: "boolean", default: false },
    fix: { type: "boolean", default: false },
  },
  allowPositionals: true,
});

const path = positionals[0] ?? ".claude-plugin/marketplace.json";
const report = verify(path, values.fix ?? false);

if (values.json) {
  console.log(JSON.stringify(report.toDict(), null, 2));
} else {
  console.log(report.toTable());
}

process.exit(report.passed() ? 0 : 1);
